Sink `suppressInliningOfRecognizedMethod` functionality to OpenJ9

0xdaryl opened the pull request:
* relocate CodeGenerator `suppressInliningOfRecognizedMethod()` functionality into
  OpenJ9 as it is exclusively specialized for OpenJ9
* relocate CodeGenerator `isMethodInAtomicLongGroup()` to OpenJ9

This is the first step in the relocation of both functions.  The corresponding
code must be removed from OMR once this PR is merged.

Signed-off-by: Daryl Maier <maier@ca.ibm.com>

andrewcraik commented:
I'm not sure this is the right direction as I believe languages other than Java have used the suppressInlining functionality - it is a generally useful mechanism. The point of the mechanism is so that code generators can opt into recognized methods that they would like to override inlining on so they can be recognized as 'intrinsics' in the code generator. The code gen specific nature of this is why it is in the code generator and not in the general inliner policies etc which are not code gen specific and use virtual inheritance for simplicity.

isMethodInAtomicLongGroup may have been used for implementation of other languages since the recognized methods have been used as a form of atomic primitive for doing memory updates. The better solution may be to factor the atomic API out in a more agnostic fashion and map the Atomics to the appropriate implementation.

0xdaryl commented:
You're right that suppressInlining functionality is useful for runtimes other than Java.  That's why I left the CodeGenerator stub in OMR (which returns false) that any downstream project can specialize as it likes (see OMR PR https://github.com/eclipse/omr/pull/1825).  The current implementations on the different processor architectures in OMR are very Java specific and it's that code that is being moved into OpenJ9.

At the moment, `isMethodInAtomicLong` is very Java-specific, checking recognized methods that map to Java juc methods.  I'd be surprised if other downstream projects have piggy-backed on that without refactoring this functionality.  While I don't disagree that a more general form would be useful, its far beyond the scope of this cleanup PR to move currently Java-specific code out of OMR.

0xdaryl commented:
@andrewcraik : have I addressed your concerns?  I did look deeper into downstream project usage of `isMethodInAtomicLongGroup` and could only find OpenJ9 as a consumer (in public repos at least).

@fjeremic @ymanton @leonardo2718 @mstoodle : for your awareness

fjeremic commented on runtime/tr.source/trj9/codegen/J9CodeGenerator.cpp:
```diff
@@ -4616,3 +4616,33 @@ J9::CodeGenerator::needRelocationsForStatics()
+         return false;
+      }
+
+   return false;
```
This return seems redundant.

0xdaryl replied:
Agreed.  Fixed.

andrewcraik commented:
@0xdaryl yes thank you - if it is not being used then it is safe to sink it down and probably a good thing for OMR so no objections.

fjeremic commented:
Jenkins test sanity all

fjeremic commented:
Jenkins sanity test all

fjeremic commented:
Jenkins test sanity

AdamBrousseau commented:
Jenkins test sanity
